### ISSUE TITLE
Interpreter does not handle ref properly

### ISSUE THREAD

**Issue by BitPuffin** [boo-lang/boo#149]
This code:

```
def Haha(ref a as int):
	a = 28

mm as int
Haha(mm)
```

Compiles just fine with booc, no errors.

However, when executing code like that in the interpreter you get an error like this:

```
ERROR: The best overload for the method 'TestModule.Haha(int)' is not compatible with the argument list '(int)'.
```

**Comment by masonwheeler** [comment]
That's very strange.  It looked like an obvious `Ducky` vs. non-`Ducky` issue, but I just tested it with `interpreter.Ducky = false` and got the same error.

I'll look at this later on today.

**Comment by BitPuffin** [comment]
@masonwheeler Do you mean that for now I can work around it by turning ducky on/off?

**Comment by masonwheeler** [comment]
No, I mean you *can't* work around it by turning ducky on/off.  That was mostly written as a note to myself: "on the surface this looks like an obvious `Ducky` problem, but it actually isn't so don't bother investigating that."

**Comment by BitPuffin** [comment]
Ah yeah sorry I should have read what you said more carefully.

**Comment by masonwheeler** [comment]
The root of the problem is in [`CallableResolutionService.CanLoadAddress`](https://github.com/boo-lang/boo/blob/master/src/Boo.Lang.Compiler/TypeSystem/Services/CallableResolutionService.cs#L176), which allows variables to be passed to a `ref` parameter if their `Entity.EntityType` is `Local`, `Parameter` or `Field`.  Because `mm` was declared in the interpreter, its entity is an [`InterpreterEntity`,](https://github.com/boo-lang/boo/blob/master/src/Boo.Lang.Interpreter/AbstractInterpreter.boo#L298) with an `EntityType` of `Custom`.

Fixing this won't be easy.  It'll take a bit of research first...

**Comment by BitPuffin** [comment]
Sounds to me then maybe that things that implement ITypedEntity should perhaps specify, or at least be able to specify, if they CanLoadAddress ?

**Comment by masonwheeler** [comment]
It goes deeper than that, actually.  

```
namespace Struct_test

import System

struct Mmm:
	public mm as int
	
	override def ToString() as string:
		return mm.ToString()

class Hmmm:
	public mm as int
	
	override def ToString() as string:
		return mm.ToString()


def Haha(ref a as int):
	a = 28

mm as int
Haha(mm)
print mm

mmm as Mmm
Haha(mmm.mm)
print mmm

hmm = Hmmm()
Haha(hmm.mm)
print hmm

mmm.mm = 5
print mmm

Console.ReadKey(true)
```

Run this as a console app, and you get `28` 3 times, followed by a 5, as expected.  Run it in the interpreter, and it's a mess.  The first time, with a local value, you get an error.  The second time, as a field on a struct, you get 0.  The third time, as a field on a class, you get 28, and the fourth time, directly assigning to a field on a local-declared struct, you still get 0.  It appears that `Boo.Lang.Interpreter` doesn't deal well with value semantics for interpreter-declared variables.

@bamboo Would you be able to take a look at this?

**Comment by BitPuffin** [comment]
Yeah I was already aware of the third one when it's a field of a class, because we worked around the other issue I had (with object conversion to bool) that somehow worked when we put it in a class called ```Box[of T]``` that just wraps a (usually struct) value in a class and we noticed that the ref code worked when we used that as well.

But yes it makes me a little uncomfortable how the semantics differ so much in the interpreter. Makes me wonder if I'll soon need to swap it out for taking the region of text and _compiling_ it and running or something, since all I'm doing is running a snippet of code..

Ideally though, Interpreter should share semantics with the compiler more or less exactly.

**Comment by BitPuffin** [comment]
Also experiencing problems with ```out``` parameters I think, probably related to this. Unless I'm just being stupid, as far as I can remember boo handles the "out" syntax part for you unlike C#, not that it matters, tried both.
